gc.c

```diff
@@ -8209,6 +8022,172 @@ gc_marks_step(rb_objspace_t *objspace, size_t slots)
 }
 #endif
 
+static bool
+gc_compact_heap_cursors_met_p(rb_heap_t *heap)
+{
+    GC_ASSERT(heap->sweeping_page && heap->compact_cursor);
+    return heap->sweeping_page == heap->compact_cursor;
+}
+
+static void
+gc_compact_sweep_unswept_page(rb_objspace_t *objspace, rb_heap_t *heap, struct heap_page *page)
+{
+    if (page->flags.before_sweep) {
+        struct gc_sweep_context ctx = {
+            .page = page,
+            .final_slots = 0,
+            .freed_slots = 0,
+            .empty_slots = 0,
+        };
+        gc_sweep_page(objspace, heap, &ctx);
+    }
+}
+
+static bool
+gc_compact_move(rb_objspace_t *objspace, rb_heap_t *heap, VALUE src)
+{
+    GC_ASSERT(BUILTIN_TYPE(src) != T_MOVED);
+    rb_heap_t *dheap = heap;
+
+    if (gc_compact_heap_cursors_met_p(dheap)) {
+        return false;
+    }
+    gc_compact_sweep_unswept_page(objspace, dheap, dheap->sweeping_page);
```

Comment:
It kind of is - that's why I checked the `before_sweep` flag inside `gc_compact_sweep_unswept_page`. It'll only be swept the first time we try and move an object into it because `gc_sweep_page` sets `before_sweep` to false.

Reply:
But why does this need to be called? Wouldn't the `try_move` call fail and then a page gets swept?

Reply:
In the current version of this code, when `try_move` fails, we increment the `sweeping_page` first and then sweep that page before retrying - so I wanted to make sure that the first time we try and move an object into the first sweeping page that that one gets swept. Or else the first page in each heap will always end up fragmented.

This part of the algorithm could probably be clearer.